crates/multi-transaction-payment/src/tests.rs

```diff
@@ -0,0 +1,417 @@
+            assert_eq!(dry_run_swapped(6, Some(6), err, Pays::Yes), baseline);
+
+            // Partial refund because less weight was used
+            let cost = dry_run_swapped(6, Some(3), err, Pays::Yes);
+            assert_eq!(cost.fee_cost, baseline.fee_cost);
+            assert!(cost.native_cost < 0); // refunded in native currency
+            assert!(cost.native_cost > refunded_baseline.native_cost); // not refunded as much as when its free
+        }
+    });
+}
+
+#[test]
+/// test that tips are paid in the fee currency, not the native currency
+fn pay_in_swapped_fee_with_tip() {
+    new_test_ext().execute_with(|| {
+        let baseline = dry_run_swapped_with_tip(6, None, false, Pays::Yes, 0);
+        assert!(baseline.fee_cost > 0); // decrease in fee currency, no change in native
+        assert_eq!(baseline.native_cost, 0);
+
+        // Paying tip increases the amount of fee taken in the fee currency
+        let cost = dry_run_swapped_with_tip(6, None, false, Pays::Yes, 1);
+        assert!(cost.fee_cost > baseline.fee_cost);
+        assert_eq!(cost.native_cost, 0);
+
+        let pays_no_baseline = dry_run_swapped_with_tip(6, None, false, Pays::No, 0);
+        assert!(pays_no_baseline.fee_cost > 0); // decrease in fee currency, no change in native
+        assert!(pays_no_baseline.native_cost < 0); // we get refunded
+
+        // `Pays::No` doesn't undo tip, so the same amount should get refunded as the case with 0 fee
+        let cost = dry_run_swapped_with_tip(6, None, false, Pays::No, 1);
+        assert!(cost.fee_cost > pays_no_baseline.fee_cost);
+        assert_eq!(cost.native_cost, pays_no_baseline.native_cost);
+    })
+}
+
+#[test]
+fn pay_fee_in_other_currency_succeeds() {
+    new_test_ext().execute_with(|| {
+        // <Payment as SignedExtension>::pre_dispatch();
+        let native_currency = GetNativeCurrencyId::get();
+        let fee_currency = GetForeignCurrencyId::get();
+        let swap_path = vec![fee_currency, native_currency];
+        setup_dex(fee_currency, native_currency);
+
+        let payment_extension = pallet_transaction_payment::ChargeTransactionPayment::<Test>::from(1);
+        let initial = 100000000000000;
+        let who = 1;
+        let call = RuntimeCall::MultiTransactionPayment(crate::Call::with_fee_swap_path {
+            path: swap_path.clone(),
+            amount_in_max: u128::MAX,
+            call: Box::new(RuntimeCall::System(frame_system::Call::remark { remark: vec![] })),
+        });
+        let weight = Weight::from_ref_time(5);
+        let info = DispatchInfo {
+            weight: weight,
+            ..Default::default()
+        };
+        let len = 123;
+
+        assert_ok!(Tokens::set_balance(
+            RuntimeOrigin::root(),
+            who,
+            fee_currency,
+            initial,
+            0
+        ));
+
+        // check that validate executes successfully, but don't commit storage
+        dry_run(|| {
```

Comment:
Could this test not be refactored to use `dry_run_given_call` if that takes an additional closure to check assertions?

Reply:
`dry_run_given_call` takes a `RuntimeCall`, i.e. an extrinsic, which `validate` isn't. And anyway I think the current solution is quite nice :)

Reply:
What I mean is that this entire test (`pay_fee_in_other_currency_succeeds`) follows almost exactly the same pattern as `dry_run_given_call`, except for checking correctness with `assert!(balance < initial);`. No big deal though